Mathlib/Analysis/Calculus/FirstDerivativeTest.lean

```diff
@@ -0,0 +1,176 @@
+    (Differentiable.differentiableAt differentiable_neg)
+    (DifferentiableOn.differentiableAt hd₁ (Ioo_mem_nhds hx.1 hx.2))) ▸ (by
+    rw [deriv_neg'', neg_mul, one_mul, Left.neg_nonpos_iff]
+    exact h₁ _ hx
+  )
+
+/-!
+### The First-Derivative Test
+
+Using the connection beetween monotonicity and derivatives we obtain the familiar
+First-Derivative Test from calculus.
+-/
+
+
+ /-- The First-Derivative Test from calculus, maxima version.
+  Suppose `a < b < c`, `f : ℝ → ℝ` is continuous at `b`,
+  the derivative `f'` is nonnegative on `(a,b)`, and
+  the derivative `f'` is nonpositive on `(b,c)`. Then `f` has a local maximum at `a`. -/
+lemma first_derivative_test_Ioo_max {f : ℝ → ℝ} {a b c : ℝ} (g₀ : a < b) (g₁ : b < c)
+    (h : ContinuousAt f b)
+    (hd₀ : DifferentiableOn ℝ f (Set.Ioo a b))
+    (hd₁ : DifferentiableOn ℝ f (Set.Ioo b c))
+    (h₀ :  ∀ x ∈ Set.Ioo a b, 0 ≤ deriv f x)
+    (h₁ :  ∀ x ∈ Set.Ioo b c, deriv f x ≤ 0) : IsLocalMax f b :=
+  have continuous_Ioc : ContinuousOn f (Ioc a b) :=
+    fun _ hx ↦ (Ioo_union_right g₀ ▸ hx).elim
+    (fun hx ↦ (hd₀.differentiableAt <| Ioo_mem_nhds hx.1 hx.2).continuousAt.continuousWithinAt)
+    (fun hx ↦ mem_singleton_iff.1 hx ▸ h.continuousWithinAt)
+  have continuous_Ico : ContinuousOn f (Ico b c) :=
+    fun _ hx ↦ (Ioo_union_left g₁ ▸ hx).elim
+    (fun hx ↦ (hd₁.differentiableAt <| Ioo_mem_nhds hx.1 hx.2).continuousAt.continuousWithinAt)
+    (fun hx ↦ mem_singleton_iff.1 hx ▸ h.continuousWithinAt)
+  isLocalMax_of_mono_anti g₀ g₁
+    (monotoneOn_of_deriv_nonneg (convex_Ioc a b)
+    continuous_Ioc (by simp_all) (by simp_all))
+    (antitoneOn_of_deriv_nonpos (convex_Ico b c)
+    continuous_Ico (by simp_all) (by simp_all))
+
+/-- The First-Derivative Test from calculus, minima version. -/
+lemma first_derivative_test_Ioo_min {f : ℝ → ℝ} {a b c : ℝ} (h : ContinuousAt f b)
+    (g₀ : a < b) (g₁ : b < c)
+    (hd₀ : DifferentiableOn ℝ f (Set.Ioo a b)) (hd₁ : DifferentiableOn ℝ f (Set.Ioo b c))
+    (h₀ : ∀ x ∈ Set.Ioo a b, deriv f x ≤ 0)
+    (h₁ : ∀ x ∈ Set.Ioo b c, 0 ≤ deriv f x) : IsLocalMin f b := by
+    have Q := @first_derivative_test_Ioo_max (-f) a b c g₀ g₁
+      (by simp_all) (DifferentiableOn.neg hd₀) (DifferentiableOn.neg hd₁)
+      (fun _ => deriv_neg_nonneg hd₀ h₀) (fun _ => deriv_neg_nonpos hd₁ h₁)
+    unfold IsLocalMax IsMaxFilter at Q
+    simp only [Pi.neg_apply, neg_le_neg_iff] at Q
+    exact Q
+
+
+ /-- The First-Derivative Test from calculus, maxima version,
+ expressed in terms of left and right filters. -/
+lemma first_derivative_test_max₀ {f : ℝ → ℝ} {b : ℝ} (h : ContinuousAt f b)
+    (hd₀ : ∀ᶠ x in 𝓝[<] b, DifferentiableAt ℝ f x) (hd₁ : ∀ᶠ x in 𝓝[>] b, DifferentiableAt ℝ f x)
+    (h₀  : ∀ᶠ x in 𝓝[<] b, 0 ≤ deriv f x) (h₁  : ∀ᶠ x in 𝓝[>] b, deriv f x ≤ 0) :
+    IsLocalMax f b := by
+  unfold Filter.Eventually at h₀ h₁ hd₀ hd₁
+  rw [mem_nhdsWithin_Iio_iff_exists_Ioo_subset] at h₀ hd₀
+  rw [mem_nhdsWithin_Ioi_iff_exists_Ioo_subset] at h₁ hd₁
```

Comment:
I was able to make that work, at the cost of adding
`change Ioo u₀ b ⊆ {x | DifferentiableAt ℝ f x} at diff_u₀`

Reply:
Do you know you can use `obtain ⟨u₀, hu₀, diff_u₀ : Ioo u₀ b ⊆ {x | DifferentiableAt ℝ f x}⟩ :=` instead of adding a `change` line? I am not saying this is necessarily the best thing to do here, I would have to see the new code to know.

Reply:
Oh, that makes sense and I do think it's an improvement.